x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java

```diff
@@ -169,6 +168,25 @@ private void resolveRoleNames(Set<String> roleNames, ActionListener<RolesRetriev
         }, listener::onFailure));
     }
 
+    private Set<RoleDescriptor> filterRolesNotUsingLicensedFeatures(Set<RoleDescriptor> roleDescriptors) {
+        final Map<Boolean, Set<RoleDescriptor>> roles = roleDescriptors.stream()
+            .collect(Collectors.partitioningBy(RoleDescriptor::isUsingDocumentOrFieldLevelSecurity, Collectors.toSet()));
+
+        final Set<RoleDescriptor> rolesUsingFeatures = roles.get(true);
+        logger.warn(
+            "User roles [{}] are disabled since they require document or field level security to determine user access. "
+                + "These security features [{}, {}] are not available under the current license. "
+                + "Users will be denied access to documents or fields granted by above roles. "
+                + "To re-enable the roles, upgrade license to [{}] or above, or renew if it's expired.",
+            DOCUMENT_LEVEL_SECURITY_FEATURE.getName(),
+            FIELD_LEVEL_SECURITY_FEATURE.getName(),
+            rolesUsingFeatures.stream().map(RoleDescriptor::getName).collect(Collectors.joining(",")),
```

Comment:
Sanity check: I'm assuming there generally won't be many roles so it's fine to include them all. Is that valid or should I add a e.g. `limit(5)` to keep this from getting too lengthy?

Reply:
Not going with the limit since there shouldn't be too many roles